Fix for issue #7413 - Allow non-zero exit codes for inspec provisioner

finchr opened the pull request:
I ran into the same problem addressed by #7413 and took a stab at fixing it.

I found and applied the fix from the PR https://github.com/jrbeilke/packer-provisioner-inspec/pull/11 to this repo and it appears to address the problem.

I will remove the WIP when I have figured out how to write a test case.

Closes #7413

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/packer?pullRequest=10723)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/packer?pullRequest=10723)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/packer?pullRequest=10723).</sub>

finchr commented:
I just pushed a new test for the valid exit codes.

rfinch-redviolet commented:
All the tests have passed, just waiting on approval. What are the next steps?

finchr commented:
I fixed a bug I found the getVersion method and pushed a fix for it and the getVersion test. I also fixed the PACKE_ACC=1 test cases so they all pass now.

azr commented on provisioner/inspec/provisioner.go:
```diff
@@ -412,9 +418,33 @@ func (p *Provisioner) executeInspec(ui packersdk.Ui, comm packersdk.Communicator
+				}
+				if !validExitCode {
+					return fmt.Errorf(
+						"Inspec exited with non-zero exit status: %d. Allowed exit codes are: %v",
```
```suggestion
						"Inspec exited with unexpected exit status: %d. Allowed exit codes are: %v",
```

mbainter replied:
Perhaps replace "Allowed exit codes" with "expected exit codes" as well, so it doesn't imply that only those codes are allowed.

rfinch-redviolet replied:
> Perhaps replace "Allowed exit codes" with "expected exit codes" as well, so it doesn't imply that only those codes are allowed.

Done, I just pushed the change.

rfinch-redviolet replied:
@azr I added your suggestion as well.

sylviamoss reviewed (commented):
This looks good! The only thing missing is the documentation. 
You should add the documentation for the new configuration in https://github.com/hashicorp/packer/blob/master/website/content/docs/provisioners/inspec.mdx

finchr commented:
> This looks good! The only thing missing is the documentation.
> You should add the documentation for the new configuration in https://github.com/hashicorp/packer/blob/master/website/content/docs/provisioners/inspec.mdx

Thanks for the link @sylviamoss, I didn't know where the documentation lived in the repo! I'll have an update in sometime today.

finchr commented:
Hi @sylviamoss I just pushed a change with documentation. I also changed the default values for `valid_exit_codes` to avoid errors with skipped inspec tests.

sylviamoss commented on provisioner/inspec/provisioner.go:
```diff
@@ -157,6 +159,10 @@ func (p *Provisioner) Prepare(raws ...interface{}) error {
 	}
 
+	if p.config.ValidExitCodes == nil {
+		p.config.ValidExitCodes = []int{0, 101}
```
https://docs.chef.io/inspec/cli/#exec leaving here some docs of why 101. 
`101  normal exit, at least one test skipped but none failed`

sylviamoss reviewed (approved):
Nice! Thanks for updating the docs. This LGTM!
